src/string_impls/to_string.rs

```diff
@@ -13,16 +13,358 @@
+    /// each value within angle brackets (`<` and `>`).
+    ///
+    /// # Warning
+    ///
+    /// When this is used, the resulting outpuot cannot then be parsed back into [`Value`]
+    /// (in part since the user can output arbitrary content for the context). Nevertheless,
+    /// writing the context can be useful for debugging errors and providing more verbose output.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use scale_value::{value, Value, ValueDef, Primitive};
+    /// use scale_value::stringify::to_writer_custom;
+    /// use std::fmt::Write;
+    ///
+    /// let val = Value {
+    ///     value: ValueDef::Primitive(Primitive::Bool(true)),
+    ///     context: "hi"
+    /// };
+    ///
+    /// let mut s = String::new();
+    ///
+    /// to_writer_custom()
+    ///     .format_context(|ctx, w: &mut &mut String| write!(w, "context: {ctx}"))
+    ///     .write(&val, &mut s)
+    ///     .unwrap();
+    ///
+    /// assert_eq!(s, r#"<context: hi> true"#);
+    /// ```
+    pub fn format_context<F: Fn(&T, &mut W) -> core::fmt::Result + 'static>(
+        mut self,
+        f: F,
+    ) -> Self {
+        self.print_context = Some(Box::new(f));
+        self
+    }
+
+    /// Add a custom formatter (for example, [`crate::stringify::custom_formatters::format_hex`]).
+    /// Custom formatters have the opportunity to read the value at each stage, and:
+    ///
+    /// - Should output `None` if they do not wish to override the standard formatting (in this case,
+    ///   they should also avoid writing anything to the provided writer).
+    /// - Should output `Some(core::fmt::Result)` if they decide to override the default formatting at
+    ///   this point.
+    ///
+    /// Custom formatters are tried in the order that they are added here, and when one decides
+    /// to write output (signalled by returning `Some(..)`), no others will be tried. Thus, the order
+    /// in which they are added is important.
+    pub fn add_custom_formatter<F: Fn(&Value<T>, &mut W) -> Option<core::fmt::Result> + 'static>(
+        mut self,
+        f: F,
+    ) -> Self {
+        self.custom_formatters.push(Box::new(f));
+        self
+    }
+
+    /// Write some value to the provided writer, using the currently configured options.
+    pub fn write(&self, value: &Value<T>, writer: W) -> core::fmt::Result {
+        let mut formatter = self.as_formatter(writer);
+        fmt_value(value, &mut formatter)
+    }
+
+    fn as_formatter(&self, writer: W) -> Formatter<'_, T, W> {
+        Formatter {
+            writer,
+            style: self.style,
+            custom_formatters: &self.custom_formatters,
+            indent_by: &self.indent_by,
+            leading_indent: &self.leading_indent,
+            print_context: &self.print_context,
+        }
+    }
+}
+
+struct Formatter<'a, T, W> {
+    writer: W,
+    style: FormatStyle,
+    custom_formatters: &'a [CustomFormatter<T, W>],
```

Comment:
nit: maybe we can have a slice of known size accounting for every type of value we can encounter? 
But this will allow only to have one custom printer per type

Reply:
Mmm, I guess I'm going for maximum flexibility here because there are times we may want to be able to custom-format the same sort of value in different ways, eg convert to hex if Composite::unnamed of Primitive::u128's that all fit into the range of a u8, or convert to ss58 if they meet some other criteria eg length is right or something :)

Reply:
Got it